### PR TITLE
Make Popup more modal.

### LINKED ISSUE #3965: Allow for capturing mouse when menus and context menus are open.
Currently when menus and context menus are open we can still partly interact with the rest of the application via pointer over effects and tooltips (and more). Most of the UI frameworks like WPF/Qt prevent such interaction as it is confusing for the user.

Ideally when context menu is open user should be able to interact with the menu itself and nothing in the main app.

### PR DESCRIPTION
**Pull request by grokys**
## What does the pull request do?

As described in #3965, mouse move etc events are still passed to the parent window when a popup is open. This doesn't happen in other frameworks, e.g. WPF.

WPF uses a capture on the mouse but that isn't feasible in Avalonia as we need a pointer event in order to capture the pointer. First attempt used event filtering but that caused problems with tooltips getting "stuck" (see comments) so this PR now uses an overlay layer:

- [x] Added `Popup.IsLightDismissEnabled` property (from UWP) and deprecate `StaysOpen` in favor of this property
- [x] Added `Popup.OverlayDismissEventPassThrough` to control whether dismiss events are passed through from the overlay layer to the underlying window content
  - The previous method of controlling this for `ComboBox` (introduced in #3857) will no longer work. Instead the `OverlayDismissEventPassThrough` property should be set on the `Popup` in the `ComboBox` template
- [x] Added `LightDismissOverlayLayer.InputPassThroughElement` which implements something similar to UWP's `FlyoutBase.OverlayInputPassThroughElement`. With this feature, `Menu`/`MenuItem` can use `IsLightDismissEnabled="True"`
  - This is needed because the top-level menus on the main window still need to receive input events even when the menu popup is open
  - Implemented by adding a `ICustomHitTest` which is similar to the existing `ICustomSimpleHitTest` but is recognized by the deferred renderer as well as the immediate renderer
  - ~~We might want to also add this to `Popup` but given that in UWP it's defined on `FlyoutBase` instead of `Popup` and that `Menu` goes direct to the overlay layer, I haven't done this in this PR~~ Added as `Popup.OverlayInputPassThroughElement`.

Also added nullable annotations to the files I touched.

## Fixed issues

Fixes #3965

### CODE REVIEW

**Comment by grokys** [comment]
@kekekeks mentioned in chat this he doesn't like the filtering raw input method of doing this, and I'm inclined to agree. If we don't use filtering, what are our options? The options that come to mind are:

1. Require that popups be opened as a result of interaction and pass an `IPointer`, we can then capture on this pointer.
  - Can't open popups without interaction (e.g. on a timer)
  - We can only capture events from the device that initiated the event
2. Add an overlay layer to the parent window/popup which to disable the window/popup contents
  - Looking at UWP, this may be the way it's done there
  - We could also implement a [`LightDismissOverlayMode`](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.primitives.popup.lightdismissoverlaymode?view=winrt-19041) like in UWP
  - Won't work if the parent window doesn't have a `VisualLayerManager`
  - Feels a bit hacky?

**Comment by kekekeks** [comment]
>Require that popups be opened as a result of interaction and pass an IPointer, we can then capture on this pointer. 

Note that this won't work with touch pointers since pointer completely disappears when touch contact is lost.

**Comment by maxkatz6** [comment]
@grokys
Not sure if I cleary understand first option. Does it means that it will depend only on pointer? "Can't open popups without interaction (e. q. on timer)" sounds like "can't open popup programmatically". 
Popup should be able to be opened from keyboard (menu button) or something like TeachingTip control, which is Popup based, could be opened on any place of screen without pointer interaction. 
Also ToolTip is always delayed with timer. 

About second options. LightDismissOverlayMode hasn't special logic to confugure dismiss mode, it is just indicator, should it be darkened. You probably need [Popup.IsLightDismissEnabled](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.primitives.popup.islightdismissenabled?view=winrt-19041#Windows_UI_Xaml_Controls_Primitives_Popup_IsLightDismissEnabled)

**Comment by grokys** [comment]
@maxkatz6 yeah I think option 1 is a no-go, so it's either the solution implemented here (raw event filtering) or option 2.

@kekekeks which do you prefer?

**Comment by grokys** [comment]
> About second option. LightDismissOverlayMode is just indicator, should background be darkened. You probably need Popup.IsLightDismissEnabled

Yeah, I realize that. I was using `LightDismissOverlayMode` as a clue that UWP does in fact add a layer to disable the window contents.

**Comment by MarchingCube** [comment]
@grokys @kekekeks Any progress on this matter? It is fairly important for user experience since otherwise we end up with random tooltips opening and pointerover effects that confuse people.

**Comment by grokys** [comment]
No heard any more feedback on this, so going to merge it as-is if it works for @MarchingCube ?

File: src/Avalonia.Controls/Primitives/Popup.cs [hunk 1]
```diff
@@ -494,13 +494,25 @@ private void CloseCore(EventArgs? closeEvent)
...
         {
-            var mouse = e as RawPointerEventArgs;
-
-            if (!StaysOpen && mouse?.Type == RawPointerEventType.NonClientLeftButtonDown)
+            if (!StaysOpen && _openState is object && e is RawPointerEventArgs pointer)
             {
-                CloseCore(e);
+                if (pointer.Root == _openState.TopLevel)
+                {
+                    // When StaysOpen == true, prevent move, wheel and leave window events being
```
Comment (grokys) [inline src/Avalonia.Controls/Primitives/Popup.cs:503]: Should be `StaysOpen == false`.

**Comment by grokys** [comment]
As @MarchingCube told me in chat: this works fine for most popups but not menus as they use `StaysOpen= "True"`.

**Comment by grokys** [comment]
Another problem with this:

- Button has a context menu and a tooltip
- Button is hovered, tooltip timer starts
- Context menu is opened
- Tooltip timer fires, tooltip is shown
- Tooltip doesn't get hidden even if mouse is no longer over the button because pointer events are being filtered out

I'm starting to think that a transparent overlay layer would be a better way to solve this...

**Review by MarchingCube — DISMISSED** [review]
Functionality wise - LGTM. Covers my use cases but maybe other people can check as well.

**Comment by MarchingCube** [comment]
Actually I will do more testing since there might be issues with pointer capture.

**Comment by MarchingCube** [comment]
Looks like that clicking on overlay with input passthrough causes controls to get stuck in pressed state.

overlay-passthrough-capture

**Review by MarchingCube — APPROVED** [review]
LGTM. Working great now.

**Pull request merged**